memory/memory.go

```diff
@@ -1,10 +1,12 @@
 package memory
 
 import (
+	"context"
 	"io"
 	"sync"
 	"time"
 
+	"golang.org/x/sync/semaphore"
```

Comment:
can we do the same with `sync.WaitGroup`

Reply:
And without it, you can just do with a channel.

```sql
n := 5 // number of workers
// first fill the channel up to N workers
var ch = make(chan struct{}, n)
for i := 0; i < n; i++ {
        ch <- struct{}{}
}

// before starting a new worker, wait until there's a spot available
<-ch

// when the worker is finish, just generate a new token
ch <- struct{}{}
```

Reply:
Yeah, we could go with a channel. Do you think it's worth it to avoid using `golang.org/x/`?